File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/setup-tableau.rb
#!/usr/bin/env ruby
# Store Tableau credentials so any coding agent can load them:
#   - ~/.claude/settings.json   — Claude Code auto-loads this into the env
#   - ~/.sigma-migration/env    — neutral, sourceable file every other agent uses
# Use this when the Tableau MCP isn't available (or when the workbook has an embedded
# datasource that the MCP can't see). See refs/tableau-rest.md.

require 'io/console'
require 'json'
require 'fileutils'

SETTINGS_PATH = File.expand_path("~/.claude/settings.json")
NEUTRAL_PATH  = File.expand_path("~/.sigma-migration/env")

# Upsert `export KEY='value'` lines into the neutral cred file (0600), preserving
# any other vars already there (e.g. Sigma creds from setup.rb).
def upsert_neutral_env(pairs)
  FileUtils.mkdir_p(File.dirname(NEUTRAL_PATH), mode: 0o700)
  body = File.exist?(NEUTRAL_PATH) ? File.read(NEUTRAL_PATH) : ""
  pairs.each do |k, v|
    line = "export #{k}='#{v}'"
    if body =~ /^export #{Regexp.escape(k)}=.*$/
      body = body.sub(/^export #{Regexp.escape(k)}=.*$/, line)
    else
      body += "\n" unless body.empty? || body.end_with?("\n")
      body += line + "\n"
    end
  end
  File.write(NEUTRAL_PATH, body)
  File.chmod(0o600, NEUTRAL_PATH)
end

puts "Tableau credential setup"
puts "Values are stored in #{SETTINGS_PATH} and loaded automatically into every Claude Code session."
puts

print "Deployment — [c]loud or self-hosted [s]erver? [c]: "
is_server = $stdin.gets.to_s.strip.downcase.start_with?("s")

if is_server
  print "Server URL (e.g. https://tableau.mycompany.com): "
  server = $stdin.gets.chomp
else
  print "Server URL [https://10ay.online.tableau.com]: "
  server = $stdin.gets.chomp
  server = "https://10ay.online.tableau.com" if server.empty?
end
server = server.sub(%r{/+$}, '')

if is_server
  print "Site contentUrl (path segment after /site/ in the URL; LEAVE BLANK for the Default site): "
else
  print "Site contentUrl (the path segment after /site/ in the Tableau URL, e.g. 'dataflow'): "
end
content_url = $stdin.gets.chomp

print "PAT name (the label you typed when creating the token in Tableau): "
pat_name = $stdin.gets.chomp

print "PAT secret (will be hidden): "
pat_secret = $stdin.noecho(&:gets).chomp
puts

# contentUrl is intentionally allowed to be empty — that's the Tableau Server
# "Default" site. Everything else is required.
if [server, pat_name, pat_secret].any?(&:empty?)
  abort "Server URL, PAT name, and PAT secret are required. Aborting without writing settings."
end

settings = File.exist?(SETTINGS_PATH) ? JSON.parse(File.read(SETTINGS_PATH)) : {}
settings["env"] ||= {}
settings["env"]["TABLEAU_SERVER_URL"]        = server
settings["env"]["TABLEAU_SITE_CONTENT_URL"]  = content_url
settings["env"]["TABLEAU_PAT_NAME"]          = pat_name
settings["env"]["TABLEAU_PAT_SECRET"]        = pat_secret

File.write(SETTINGS_PATH, JSON.pretty_generate(settings))

upsert_neutral_env(
  "TABLEAU_SERVER_URL"       => server,
  "TABLEAU_SITE_CONTENT_URL" => content_url,
  "TABLEAU_PAT_NAME"         => pat_name,
  "TABLEAU_PAT_SECRET"       => pat_secret,
)

puts "Wrote Tableau credentials to:"
puts "  #{SETTINGS_PATH}  (Claude Code auto-loads this)"
puts "  #{NEUTRAL_PATH}  (any other agent / shell)"
puts
puts "Claude Code: open a new session (or `! source ~/.claude/settings.json`)."
puts "Other agents / shell: `source ~/.sigma-migration/env`, then run"
puts "`eval \"$(scripts/get-tableau-token.sh)\"` to sign in."

File: plugins/tableau-to-sigma/skills/tableau-to-sigma/scripts/setup-tableau.py
#!/usr/bin/env python3
"""Persist Tableau PAT credentials without requiring Ruby."""

from __future__ import annotations

import argparse
import getpass
import json
import os
import re
import sys
import urllib.parse
from pathlib import Path
from typing import NoReturn

SETTINGS_PATH = Path.home() / ".claude" / "settings.json"
NEUTRAL_PATH = Path.home() / ".sigma-migration" / "env"
DEFAULT_CLOUD = "https://us-west-2b.online.tableau.com"


def die(message: str) -> NoReturn:
    print(message, file=sys.stderr)
    raise SystemExit(2)


def shell_single_quote(value: str) -> str:
    return "'" + value.replace("'", "'\"'\"'") + "'"


def upsert_neutral_env(pairs: dict[str, str]) -> None:
    NEUTRAL_PATH.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    body = NEUTRAL_PATH.read_text(encoding="utf-8") if NEUTRAL_PATH.exists() else ""
    for key, value in pairs.items():
        line = f"export {key}={shell_single_quote(value)}"
        pattern = re.compile(rf"^export {re.escape(key)}=.*$", re.MULTILINE)
        if pattern.search(body):
            body = pattern.sub(line, body, count=1)
        else:
            if body and not body.endswith("\n"):
                body += "\n"
            body += line + "\n"
    NEUTRAL_PATH.write_text(body, encoding="utf-8")
    try:
        NEUTRAL_PATH.chmod(0o600)
    except OSError:
        pass


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--from-env", action="store_true")
    args = parser.parse_args()

    use_env = args.from_env or not sys.stdin.isatty()
    if use_env:
        server = os.environ.get("TABLEAU_SERVER_URL", "")
        site = os.environ.get("TABLEAU_SITE_CONTENT_URL", "")
        name = os.environ.get("TABLEAU_PAT_NAME", "")
        secret = os.environ.get("TABLEAU_PAT_SECRET", "")
        missing = [
            key
            for key, value in (
                ("TABLEAU_SERVER_URL", server),
                ("TABLEAU_PAT_NAME", name),
                ("TABLEAU_PAT_SECRET", secret),
            )
            if not value
        ]
        if missing:
            die(
                "ERROR: missing environment variable(s): "
                + ", ".join(missing)
                + ". Export them and run `python scripts/setup-tableau.py --from-env`."
            )
    else:
        server = input(f"Server URL [{DEFAULT_CLOUD}]: ").strip() or DEFAULT_CLOUD
        site = input(
            "Site contentUrl (blank for the Default site): "
        ).strip()
        name = input("PAT name: ").strip()
        secret = getpass.getpass("PAT secret (hidden): ").strip()

    server = server.rstrip("/")
    parsed = urllib.parse.urlparse(server)
    if parsed.scheme != "https" or not parsed.hostname:
        die("TABLEAU_SERVER_URL must be an absolute https:// URL.")
    if not name or not secret:
        die("Server URL, PAT name, and PAT secret are required.")

    if SETTINGS_PATH.exists():
        try:
            settings = json.loads(SETTINGS_PATH.read_text(encoding="utf-8-sig"))
        except (OSError, ValueError) as exc:
            die(f"Cannot parse {SETTINGS_PATH}: {exc}")
    else:
        settings = {}
    settings.setdefault("env", {})
    pairs = {
        "TABLEAU_SERVER_URL": server,
        "TABLEAU_SITE_CONTENT_URL": site,
        "TABLEAU_PAT_NAME": name,
        "TABLEAU_PAT_SECRET": secret,
    }
    settings["env"].update(pairs)
    SETTINGS_PATH.parent.mkdir(parents=True, exist_ok=True)
    SETTINGS_PATH.write_text(
        json.dumps(settings, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    upsert_neutral_env(pairs)

    print(f"Tableau credentials saved to {SETTINGS_PATH} and {NEUTRAL_PATH}.")
    print(f"TABLEAU_SERVER_URL: {server}")
    print(f"TABLEAU_SITE_CONTENT_URL: {site or '(Default site)'}")
    print(f"TABLEAU_PAT_NAME: {name}")
    print(f"TABLEAU_PAT_SECRET: {'*' * max(0, len(secret) - 4) + secret[-4:]}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
